### PR TITLE
feat: Skip test folder name for ptn

### LINKED ISSUE #1470: Update Pester: Skip defaults|waf-aligned test folder check for ptn modules
The following Pester tests are not relevant for pattern modules, thus they should not apply to modules of type `ptn`

- Module should contain a [` tests/e2e/*defaults `] folder
- Module should contain a [` tests/e2e/*waf-aligned `] folder

This issue is about updating Pester tests to comply with the above

### PR DESCRIPTION
**Pull request by eriqua**
## Description

Closes #1470 

- If the module is of type `ptn`, skip Pester test checking for `defaults` and `waf-aligned` test folders
- Updating workflow job conditions accordingly.

PsRule jobs running only on `defaults `and `waf-aligned` folders, need to be skipped if the value of `psRuleModuleTestFilePaths `is empty. `psRuleModuleTestFilePaths `is the variable calculated by the initialize pipeline job, hosting the path to the deploy test 

Test performed in a different branch (main) where a "fake" pattern module is created with no `defaults` or `waf-aligned` test folders. Pester tests for telemetry and orphaned modules have been commented during testing. PR 1620 contains, for reference, also the "fake" files used to test this update.

1. push to main - complete, success (publish runs)
image

1. manual run - only static, success (publish is skipped)
image

1. manual run - only deployment (publish is skipped)
image

1. push to main - complete, static fails (deployment and publish are skipped)
image

1. push to main - cancel pipeline while running static (deployment and publish are skipped)
image

1. manual run - complete, success (publish runs)
image

## Pipeline Reference

| Pipeline |
| -------- |
| [avm.ptn.batch.batch-account-ptn](https://github.com/eriqua/bicep-registry-modules/actions/workflows/avm.ptn.batch.batch-account-ptn.yml) (test ptn) |
| [avm.res.key-vault.vault](https://github.com/eriqua/bicep-registry-modules/actions/workflows/avm.res.key-vault.vault.yml) (test res) |

## Type of Change

- [x] Update to CI Environment or utlities (Non-module effecting changes)
- [ ] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [ ] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1 [hunk 1]
```diff
@@ -59,7 +59,7 @@ Describe 'File/folder tests' -Tag 'Modules' {
 
     $moduleFolderTestCases = [System.Collections.ArrayList] @()
     foreach ($moduleFolderPath in $moduleFolderPaths) {
-      $resourceTypeIdentifier = ($moduleFolderPath -split '[\/|\\]{1}avm[\/|\\]{1}(res|ptn)[\/|\\]{1}')[2] -replace '\\', '/' # avm/res/<provider>/<resourceType>
+      $resourceTypeIdentifier = ($moduleFolderPath -split '[\/|\\]{1}avm[\/|\\]{1}(res|ptn)[\/|\\]{1}')[2] -replace '\\', '/' # <provider>/<resourceType>
```
Comment (AlexanderSehr) [inline avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1:62]:
```suggestion
      $resourceTypeIdentifier = ($moduleFolderPath -split '[\/|\\]{1}avm[\/|\\]{1}(res|ptn)[\/|\\]{1}')[2] -replace '\\', '/' # avm/res|ptn/<provider>/<resourceType>
```

Don't think we should just remove the example 😄

File: avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1 [hunk 2]
```diff
@@ -146,11 +146,14 @@ Describe 'File/folder tests' -Tag 'Modules' {
 
     $topLevelModuleTestCases = [System.Collections.ArrayList]@()
     foreach ($moduleFolderPath in $moduleFolderPaths) {
-      $resourceTypeIdentifier = ($moduleFolderPath -split '[\/|\\]{1}avm[\/|\\]{1}(res|ptn)[\/|\\]{1}')[2] -replace '\\', '/' # avm/res/<provider>/<resourceType>
+      Write-Verbose ("moduleFolderPath: $moduleFolderPath") -Verbose
+      $resourceTypeIdentifier = ($moduleFolderPath -split '[\/|\\]{1}avm[\/|\\]{1}(res|ptn)[\/|\\]{1}')[2] -replace '\\', '/' # <provider>/<resourceType>
```
Comment (AlexanderSehr) [inline avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1:150]: Same comment for all the other examples

**Review by AlexanderSehr — COMMENTED** [review]
Hey @eriqua, looks good. Just the one comment regarding the example

File: avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1 (hunk 1 above)
Comment (eriqua) [reply avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1:62]: That's not the case. The value of `$resourceTypeIdentifier` is `<provider>/<resourceType>` and not `avm/res|ptn/<provider>/<resourceType>` due to the `split[2]`

File: avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1 (hunk 2 above)
Comment (eriqua) [reply avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1:150]: Same reply 😄

File: avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1 (hunk 1 above)
Comment (AlexanderSehr) [reply avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1:62]: No no - it's an example what it matches against, not what the outcome is. When I'm seeing 
`[\/|\\]{1}avm[\/|\\]{1}(res|ptn)[\/|\\]{1}'`

I may we wondering what and actually string would look that the regex applies to. When I then get
`avm/res|ptn/<provider>/<resourceType>`

It helps me piece together: Ah ok a slash in the beginning, then AVM, then it looks for res or ptn, and finally another slash. With the string above that would mean if it splits by this value, what I get is ... `<provider>/<resourceType>`

Comment (AlexanderSehr) [reply avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1:62]: At least that's how I'm thinking. Which is why I added the examples in the first place :D

**Comment by ChrisSidebotham** [comment]
Was there a decision for PSRule to not run against pattern module tests?

**Comment by AlexanderSehr** [comment]
> Was there a decision for PSRule to not run against pattern module tests?
> 
> My expectation is that the singular test for patterns should also include tests for WAF Allignment and PSRule. As these patterns should be complete 'Packages'

There was, yes. I think it came out of the discussion about the Subscription Vending pattern

File: avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1 (hunk 1 above)
Comment (eriqua) [reply avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1:62]: If I see a comment next to an assignment I take for granted it is representing the value of the assignment, hence the confusion. Instead the comment refers to the regex and how it maps to the moduleFolderPath value, correct?

**Comment by eriqua** [comment]
> Was there a decision for PSRule to not run against pattern module tests?
> 
> My expectation is that the singular test for patterns should also include tests for WAF Allignment and PSRule. As these patterns should be complete 'Packages'

This may be useful indeed, but for the moment we don't have a clear direction on that, just the fact we don't want to enforce the same test folder naming as Alex shared above. I'd suggest to first merge this since patterns are currently blocked by that. Next, we should discuss if running PSRule on all ptn tests for example, whatever the test folder name is. Let's discuss this second point offline, good call

File: avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1 (hunk 1 above)
Comment (AlexanderSehr) [reply avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1:62]: Hey @eriqua , it shows how a 'moduleFolderPath' would look like. We could also change it to `# 'avm/res|ptn/<provider>/<resourceType>' would return '<provider>/<resourceType>'`

Comment (eriqua) [reply avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1]: Done @AlexanderSehr

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**